sqlstats: mutex improvements

xinhaoz opened the pull request:
Previously, when attempting to get a stmt or txn entry in the sql stats containers, we pass a flag to the get* functions specifying whether or not to create the entry if it does not exist. This commit cleans up this interface by splitting the function up into a pure get or tryCreate functions.

Note that originally we were pursuing a double checked lock approach, however Golang's RWMutex does not scale well with the number of CPUs. Instead this patch just simplifies the code here a bit to make some incoming changes that will eventually move this stats recording step out of the execution path easier. Once that happens we can investigate further improvment or removal of these mutexes.

Epic: none
Release note: None
Part of: #141024

### sqlstats: convert RWMutex to Mutex

This commit changes the RWMutex on the sql stats
containers to the regular exclusive Mutex struct.
The motivation behind this change is that RWMutex
scales poorly with CPU count. See benchmarks below.

Epic: none
Part of: https://github.com/cockroachdb/cockroach/issues/140590

xinhaoz committed:
sqlstats: split get{Stmt,Txn} functions up into read or write

Previously, when attempting to get a stmt or txn entry
in the sql stats containers, we pass a flag to the get*
functions specifying whether or not to create the entry if
it does not exist. This commit cleans up this interface by
splitting the function up into a pure get or tryCreate functions.

Note that originally we were pursuing a double checked lock
approach, however Golang's RWMutex does not scale well with
the number of CPUs. Instead this patch just simplifies the
code here a bit to make some incoming changes that will eventually
move this stats recording step out of the execution path easier.
Once that happens we can investigate further improvment or
removal of these mutexes.

Epic: none
Release note: None
Part of: #141024

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141045)
<!-- Reviewable:end -->

xinhaoz committed:
sqlstats: convert RWMutex to Mutex

This commit changes the RWMutex on the sql stats
containers to the regular exclusive Mutex struct.
The motivation behind this change is that RWMutex
scales poorly with CPU count. See benchmarks below.

Epic: none
Part of: #140590

Release note: None

herkolategan commented:
@xinhaoz If you have a chance do you mind rebasing this PR on the latest master, think it would be a good one to test out the microbenchmark CI pipeline (which will only run if rebased on the latest master) - would be much appreciated!

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 3 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141045#-:-OIq7dXW22z1yYwr1O8m:bg7igkv)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @xinhaoz)

___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_writer.go` line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141045#-OIq6wQ2CSpnrNsjfvb3:-OIq6wQ2CSpnrNsjfvb4:b-kybwtk) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc0abfb5d4abbf8e6085b3c408195ba40e09ca1f/pkg/sql/sqlstats/ssmemstorage/ss_mem_writer.go#L129)):*
> ```Go
> 	stats.mu.distSQLUsed = key.DistSQL
> 	stats.mu.fullScan = value.FullScan
> 	stats.mu.database = value.Database
> ```

why did these lines need to change?


<!-- Sent from Reviewable.io -->

xinhaoz commented:
I would have liked to run benchdiff with less cpus but it doesn't seem like benchdiff supports that yet. Might have to do some more investigation there.

xinhaoz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/141045)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[`pkg/sql/sqlstats/ssmemstorage/ss_mem_writer.go` line 129 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/141045#-OIq6wQ2CSpnrNsjfvb3:-OIqATUy8d2mvGGAx-HZ:b341nz7) ([raw file](https://github.com/cockroachdb/cockroach/blob/dc0abfb5d4abbf8e6085b3c408195ba40e09ca1f/pkg/sql/sqlstats/ssmemstorage/ss_mem_writer.go#L129)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

why did these lines need to change?
</blockquote></details>

They don't need to, it seems like this just leaked from another diff I had. We can keep them though because i plan to get rid of passing the key here and just have everything in `RecordedStmtStats`.


<!-- Sent from Reviewable.io -->

xinhaoz commented:
tftr!
bors r+

tbg commented:
Thanks for jumping on this @xinhaoz!
